In [1]:
import pandas as pd

df = pd.read_csv('../data/breast_cancer.csv')

X = df.drop('target', axis=1)
y = df['target']

In [2]:
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from xgboost import XGBClassifier

In [3]:
svm_pipeline = Pipeline([
    ('scaler', StandardScaler()),
    ('model', SVC())
])

rf_model = RandomForestClassifier(random_state=42)

xgb_model = XGBClassifier(eval_metric='logloss')

In [4]:
rf_scores = cross_val_score(rf_model, X, y, cv=5, scoring='accuracy')
svm_scores = cross_val_score(svm_pipeline, X, y, cv=5, scoring='accuracy')
xgb_scores = cross_val_score(xgb_model, X, y, cv=5, scoring='accuracy')

In [5]:
import numpy as np

print(f"Random Forest: {rf_scores.mean():.4f} ± {rf_scores.std():.4f}")
print(f"SVM: {svm_scores.mean():.4f} ± {svm_scores.std():.4f}")
print(f"XGBoost: {xgb_scores.mean():.4f} ± {xgb_scores.std():.4f}")

Random Forest: 0.9561 ± 0.0228
SVM: 0.9736 ± 0.0147
XGBoost: 0.9736 ± 0.0147


## 🔁 Cross-Validation Results (5-Fold)

To ensure robustness and generalization, 5-fold cross-validation was applied to all models.

### 📊 Results:
- Random Forest: **0.9561 ± 0.0228**
- SVM: **0.9736 ± 0.0147**
- XGBoost: **0.9736 ± 0.0147**

---

### 📈 Key Insights:

- **SVM and XGBoost achieved the highest mean accuracy (97.36%)**, indicating strong predictive performance
- Both models also show **low standard deviation (~0.0147)**, suggesting stable performance across different folds
- Random Forest performed slightly lower but still maintained solid generalization capability
- Low variance across all models indicates that the dataset is well-structured and suitable for classification tasks

---

### 🎯 Conclusion:

Cross-validation confirms that the models generalize well and are not dependent on a single train-test split.  
SVM and XGBoost emerge as the most reliable models in terms of both accuracy and stability.